core/src/test/java/org/elasticsearch/bootstrap/ElasticsearchCommandLineParsingTests.java

```diff
@@ -0,0 +1,195 @@
+        runDaemonizeTest(false);
+    }
+
+    private void runDaemonizeTest(final boolean expectedDaemonize, final String... args) throws Exception {
+        runTest(
+                ExitCodes.OK,
+                true,
+                output -> {},
+                (daemonize, pathHome, pidFile, esSettings) -> assertThat(daemonize, equalTo(expectedDaemonize)),
+                args);
+    }
+
+    public void testElasticsearchSettings() throws Exception {
+        runTest(
+                ExitCodes.OK,
+                true,
+                output -> {},
+                (daemonize, pathHome, pidFile, esSettings) -> {
+                    assertThat(esSettings.size(), equalTo(2));
+                    assertThat(esSettings, hasEntry("es.foo", "bar"));
+                    assertThat(esSettings, hasEntry("es.baz", "qux"));
+                },
+                "-Ees.foo=bar", "-E", "es.baz=qux"
+        );
+    }
+
+    public void testElasticsearchSettingPrefix() throws Exception {
+        runElasticsearchSettingPrefixTest("-E", "foo");
+        runElasticsearchSettingPrefixTest("-E", "foo=bar");
+        runElasticsearchSettingPrefixTest("-E", "=bar");
+    }
+
+    private void runElasticsearchSettingPrefixTest(String... args) throws Exception {
+        runTest(
+                ExitCodes.USAGE,
+                false,
+                output -> assertThat(output, containsString("Elasticsearch settings must be prefixed with [es.] but was [")),
+                (daemonize, pathHome, pidFile, esSettings) -> {},
+                args
+        );
+    }
+
+    public void testElasticsearchSettingCanNotBeEmpty() throws Exception {
+        runTest(
+                ExitCodes.USAGE,
+                false,
+                output -> assertThat(output, containsString("Elasticsearch setting [es.foo] must not be empty")),
+                (daemonize, pathHome, pidFile, esSettings) -> {},
+                "-E", "es.foo="
+        );
+    }
+
+    public void testUnknownOption() throws Exception {
+        runTest(
+                ExitCodes.USAGE,
+                false,
+                output -> assertThat(output, containsString("network.host is not a recognized option")),
+                (daemonize, pathHome, pidFile, esSettings) -> {},
+                "--network.host");
+    }
+
+    private interface InitConsumer {
+        void accept(final boolean daemonize, final String pathHome, final String pidFile, final Map<String, String> esSettings);
+    }
+
+    private void runTest(
+            final int expectedStatus,
+            final boolean expectedInit,
+            final Consumer<String> outputConsumer,
+            final InitConsumer initConsumer,
+            String... args) throws Exception {
+        final MockTerminal terminal = new MockTerminal();
+        try {
+            final AtomicBoolean init = new AtomicBoolean();
+            final int status = Elasticsearch.main(args, new Elasticsearch() {
```

Comment:
That will not work because the new instance closes over a local variable and a method parameter.